Modules/_io/fileio.c

```diff
@@ -519,9 +542,19 @@ new_buffersize(fileio *self, size_t currentsize)
 #ifdef HAVE_FSTAT
     off_t pos, end;
     struct stat st;
-    if (fstat(self->fd, &st) == 0) {
+    int res;
+
+    Py_BEGIN_ALLOW_THREADS
+    res = fstat(self->fd, &st);
+    Py_END_ALLOW_THREADS
```

Comment:
Hi,

wouldn't it be slightly better to make the 2 sys calls in the same _ALLOW_THREADS_ block, given they follow each other in the usual case ?

like this:

```
Py_BEGIN_ALLOW_THREADS
res = fstat(self->fd, &st);
if (res == 0) {
    pos = lseek(self->fd, 0L, SEEK_CUR);
}
Py_END_ALLOW_THREADS
if (res == 0) {
    end = st.st_size;
    # etc..
```

regards,

gst.

Reply:
Thanks for the review @gst. The code is less clear this way, and we are not calling this in a tight loop, so I don't think this optimization is needed.

Reply:
agreed.
but now I see that the lseek call isn't checked for error (return value == -1)..
shouldn't it be actually ?

Reply:
I think the current error handling is fragile. But I want to make the minimal change needed  for this issue.

Reply:
The best would be to backport _Py_fstat() from master: it returns the file size properly on Windows by using GetFileInformationByHandle(). Be honestly, I don't think that it's worth it and it's perfectly fine to release/acquire the GIL twice in this short function.